File: int-st.vb
'Ed Davis. Tiny Basic that can play Star Trek

imports system.console
imports system.io
imports System.Collections

module TinyBasic

const c_maxlines = 7000, c_maxvars = 26, c_at_max = 1000

dim pgm(c_maxlines) as string         ' program stored here
dim vars(c_maxvars) as long           ' variable store
dim atarry(c_at_max) as long          ' the @ array

dim gstackln As Stack = New Stack()   ' gosub line stack
dim gstacktp As Stack = New Stack()   ' gosub textp stack

dim forvar(c_maxvars) as long
dim forlimit(c_maxvars) as long
dim forline(c_maxvars) as integer
dim forpos(c_maxvars) as integer

dim tok as string, toktype as string    ' current token, and it's type
dim thelin as string, thech as string   ' current program line, current character
dim curline as integer, textp as integer ' position in current line
dim num as long      ' last number read by scanner
dim errors as boolean
dim tracing as boolean

dim timestart as double

sub main()
  newstmt()
  help()
  do
    errors = false
    write("vb> ")
    pgm(0) = ReadLine()
    if pgm(0) <> "" then
      initlex(0)
      if toktype = "number" then
        validlinenum()
        pgm(CInt(num)) = mid(pgm(0), textp)
      else
        docmd()
      end if
    end if
  loop
end sub

sub docmd()
  dim running as boolean
  running = false
  do
    if tracing and left(tok, 1) <> ":" then writeline("[" & curline & "] " & tok & " " & thech & " " & mid(thelin, textp))
    select case tok
      case "bye", "quit" : nexttok(): end
      case "end", "stop" : nexttok(): exit sub
      case "clear"       : nexttok(): clearvars(): exit sub
      case "help"        : nexttok(): help():      exit sub
      case "list"        : nexttok(): liststmt():  exit sub
      case "load"        : nexttok(): loadstmt():  exit sub
      case "new"         : nexttok(): newstmt():   exit sub
      case "run"         : nexttok(): runstmt(): running = true
      case "save"        : nexttok(): savestmt():  exit sub
      case "tron"        : nexttok(): tracing = true
      case "troff"       : nexttok(): tracing = false
      case "cls"         : nexttok(): clear()
      case "for"         : nexttok(): forstmt()
      case "gosub"       : nexttok(): gosubstmt()
      case "goto"        : nexttok(): gotostmt()
      case "if"          : nexttok(): ifstmt()
      case "input"       : nexttok(): inputstmt()
      case "next"        : nexttok(): nextstmt()
      case "print", "?"  : nexttok(): printstmt()
      case "return"      : nexttok(): returnstmt()
      case "@"           : nexttok(): arrassn()
      case ":"           : nexttok()  ' just continue
      case ""            : ' handled below
      case else
        if tok = "let" then nexttok()
        if toktype = "ident" then
          assign()
        else
          writeline("Unknown token " & tok & " at line " & curline): errors = true
        end if
    end select

    if errors then exit sub
    if curline > c_maxlines then showtime(running): exit sub
    do while tok = ""
      if curline = 0 or curline >= c_maxlines then showtime(running): exit sub
      initlex(curline + 1)
    loop
  loop
end sub

sub showtime(running as boolean)
  if running then writeline("Took : " & Microsoft.VisualBasic.Timer - timestart & " seconds")
end sub

sub help
   writeline("+---------------------- Tiny Basic (VB.NET) --------------------------+")
   writeline("| bye, clear, cls, end/stop, help, list, load/save, new, run, tron/off|")
   writeline("| for <var> = <expr1> to <expr2> ... next <var>                       |")
   writeline("| gosub <expr> ... return                                             |")
   writeline("| goto <expr>                                                         |")
   writeline("| if <expr> then <statement>                                          |")
   writeline("| input [prompt,] <var>                                               |")
   writeline("| <var>=<expr>                                                        |")
   writeline("| print <expr|string>[,<expr|string>][;]                              |")
   writeline("| rem <anystring>                                                     |")
   writeline("| Operators: ^, * / \ mod + - < <= > >= = <>, not, and, or            |")
   writeline("| Integer variables a..z, and array @(expr)                           |")
   writeline("| Functions: abs(expr), asc(ch), rnd(expr), sgn(expr)                 |")
   writeline("+---------------------- Tiny Basic Help ------------------------------+")
end sub

sub gosubstmt()   ' for gosub: save the line and column
  gstackln.Push(curline)
  gstacktp.Push(textp)

  gotostmt()
end sub

sub assign()
  dim var as integer
  var = getvarindex(): nexttok()
  expect("=")
  vars(var) = expression(0)
  if tracing then writeline("*** " & chr(var + asc("a")) & " = " & vars(var))
end sub

sub arrassn()   ' array assignment: @(expr) = expr
  dim n as long, atndx as integer

  atndx = CInt(parenexpr())
  if tok <> "=" then
    writeline("Array Assign: Expecting '=', found: " & tok): errors = true
  else
    nexttok()     ' skip the "="
    n = expression(0)
    atarry(atndx) = n
    if tracing then writeline("*** @(" & atndx & ") = " & n)
  end if
end sub

sub forstmt() ' for i = expr to expr
  dim var, forndx as integer, n as long
  var = getvarindex()
  assign()
  ' vars(var) has the value; var has the number value of the variable in 0..25
  forndx = var
  forvar(forndx) = vars(var)
  if tok <> "to" then
    writeline("For: Expecting 'to', found:" & tok): errors = true
  else
    nexttok()
    n = expression(0)
    forlimit(forndx) = n
    ' need to store iter, limit, line, and col
    forline(forndx) = curline
    if tok = "" then forpos(forndx) = textp else forpos(forndx) = textp - 2
  end if
end sub

sub ifstmt()
  if expression(0) = 0 then skiptoeol(): exit sub
  if tok = "then" then nexttok()      ' "then" is optional
  if toktype = "number" then gotostmt()
end sub

sub inputstmt()   ' "input" [string ","] var
  dim var as integer
  dim st as string
  if toktype = "string" then
    write(mid(tok, 2))
    nexttok()
    expect(",")
  else
    write("? ")
  end if
  var = getvarindex: nexttok()
  st = readline()
  if st = "" then st = "0"
  if left(st, 1) >= "0" and left(st, 1) <= "9" then
    vars(var) = CLng(st)
  else
    vars(var) = asc(st) ' turn characters into their ascii value
  end if
end sub

sub liststmt()
  dim i as integer
  for i = 1 to c_maxlines
    if pgm(i) <> "" then writeline(i & " " & pgm(i))
  next i
  writeline("")
end sub

sub loadstmt()
  dim n as integer, filename as string
  dim f as StreamReader

  newstmt()
  filename = getfilename("Load")
  if filename = "" then exit sub
  f = new StreamReader(filename)
  n = 0
  do until f.peek = -1
    pgm(0) = f.ReadLine()
    initlex(0)
    if toktype = "number" and num > 0 and num <= c_maxlines then
      pgm(CInt(num)) = mid(pgm(0), textp)
      n = CInt(num)
    else
      n = n + 1
      pgm(n) = pgm(0)
    end if
  loop
  f.Close()
  curline = 0
end sub

sub newstmt()
  dim i as integer
  clearvars()
  for i = 1 to c_maxlines
    pgm(i) = ""
  next i
end sub

sub nextstmt ' next ident
  dim forndx as integer

  ' tok needs to have the variable
  forndx = getvarindex()
  forvar(forndx) = forvar(forndx) + 1
  vars(forndx) = forvar(forndx)
  if tracing then writeline("*** " & chr(forndx + asc("a")) & " = " & vars(forndx))
  if forvar(forndx) <= forlimit(forndx) then
    curline = forline(forndx)
    textp   = forpos(forndx)
    'print "nextstmt tok>"; tok; " textp>"; textp; " >"; mid$(thelin, textp)
    call initlex2
  else
    call nexttok ' skip the ident for now
  end if
end sub

sub printstmt()
  dim printnl as boolean, printwidth as integer, n as long, junk as string

  printnl = true
  do while tok <> ":" and tok <> "" and tok <> "else"
    printnl = true
    printwidth = 0
    if accept("#") then
      if num <= 0 then writeline("Expecting a print width, found:" + tok): exit sub
      printwidth = CInt(num)
      nexttok()
      if not accept(",") then writeline("Print: Expecting a ',', found:" + tok): exit sub
    end if

    if toktype = "string" then
      junk = mid(tok, 2)
      nexttok()
    else
      n = expression(0)
      junk = ltrim(str(n))
    end if
    printwidth = printwidth - len(junk)
    if printwidth <= 0 then write(junk) else write(space(printwidth) + junk)

    if accept(",") or accept(";") then printnl = false else exit do
  loop

  if printnl then writeline()
end sub

sub returnstmt()    ' return from a subroutine
  curline = CInt(gstackln.Pop())
  textp   = CInt(gstacktp.Pop())
  initlex2()
end sub

sub runstmt()
  timestart = Microsoft.VisualBasic.Timer
  clearvars()
  initlex(1)
end sub

sub gotostmt()
  num = expression(0)
  validlinenum()
  initlex(CInt(num))
end sub

sub savestmt()
  dim i as integer, filename as string
  dim f as StreamWriter

  filename = getfilename("Save")
  if filename = "" then exit sub
  f = new StreamWriter(filename, false)
  for i = 1 to c_maxlines
    if pgm(i) <> "" then f.writeline(i & " " & pgm(i))
  next i
  f.close()
end sub

function getfilename(action as string) as string
  dim filename as string
  if toktype = "string" then
    filename = mid(tok, 2)
  else
    write(action & ": ")
    filename = readline()
  end if
  if filename <> "" then
    if instr(filename, ".") = 0 then filename = filename + ".bas"
  end if
  return filename
end function

sub validlinenum()
  if num <= 0 or num > c_maxlines then writeline("Line number out of range"): errors = true
end sub

sub clearvars()
  dim i as integer
  for i = 1 to c_maxvars
    vars(i) = 0
  next i

  gstackln.Clear()
  gstacktp.Clear()
end sub

function parenexpr() as long
  dim n as long

  expect("("): if errors then return 0
  n = expression(0)
  expect(")")
  return n
end function

'1  Or
'2  And
'3  Not
'4  = <> < <= > >=
'5  + -
'6  * / \ Mod
'7  - + (unary)
'8  ^
function expression(minprec as integer) as long
  dim n as long

  ' handle numeric operands - numbers and unary operators
  if toktype = "number" then
    n = num: nexttok()
  elseif tok = "-" then
    nexttok(): n = -expression(7)
  elseif tok = "+" then
    nexttok(): n =  expression(7)
  elseif tok = "not" then
    nexttok(): n = not expression(3)
  elseif tok = "abs" then
    nexttok(): n = Math.abs(parenexpr())
  elseif tok = "asc" then
    nexttok(): expect("("): n = asc(mid(tok, 2, 1)): nexttok(): expect(")")
  elseif tok = "rnd" or tok = "irnd" then
    nexttok(): n = Convert.ToInt32(rnd * parenexpr() + 1)
  elseif tok = "sgn" then
    nexttok(): n = Math.sign(parenexpr())
  elseif toktype = "ident"  then
    n = vars(getvarindex()): nexttok()
  elseif tok = "@"  then
    nexttok(): n = atarry(CInt(parenexpr()))
  elseif tok = "("  then
    n =  parenexpr()
  else
    writeline("syntax error: expecting an operand, found: " & tok): errors = true: return 0
  end if

  do  ' while binary operator and precedence of tok >= minprec
    if     minprec <= 1 and tok = "or" then
      nexttok(): n = Convert.ToInt32(n or expression(2))
    elseif minprec <= 2 and tok = "and" then
      nexttok(): n = Convert.ToInt32(n and expression(3))
    elseif minprec <= 4 and tok = "=" then
      nexttok(): n = Convert.ToInt32(n = expression(5))
    elseif minprec <= 4 and tok = "<" then
      nexttok(): n = Convert.ToInt32(n < expression(5))
    elseif minprec <= 4 and tok = ">" then
      nexttok(): n = Convert.ToInt32(n > expression(5))
    elseif minprec <= 4 and tok = "<>" then
      nexttok(): n = Convert.ToInt32(n <> expression(5))
    elseif minprec <= 4 and tok = "<=" then
      nexttok(): n = Convert.ToInt32(n <= expression(5))
    elseif minprec <= 4 and tok = ">=" then
      nexttok(): n = Convert.ToInt32(n >= expression(5))
    elseif minprec <= 5 and tok = "+" then
      nexttok(): n = n + expression(6)
    elseif minprec <= 5 and tok = "-" then
      nexttok(): n = n - expression(6)
    elseif minprec <= 6 and tok = "*" then
      nexttok(): n = n * expression(7)
    elseif minprec <= 6 and (tok = "/" or tok = "\") then
      nexttok(): n = n \ expression(7)
    elseif minprec <= 6 and tok = "mod" then
      nexttok(): n = n mod expression(7)
    elseif minprec <= 8 and tok = "^" then
      nexttok(): n = CLng(n ^ expression(9))
    else
      exit do
    end if
  loop

  return n
end function

function getvarindex() as integer
  if toktype <> "ident" then writeline("Not a variable:" & tok): errors = true: return 0
  return asc(left(tok, 1)) - asc("a")
end function

sub expect(s as string)
  if accept(s) then exit sub
  writeline("(" & curline & ") expecting " & s & " but found " & tok & " =>" & pgm(curline)): errors = true
end sub

function accept(s as string) as boolean
  if tok = s then nexttok(): return true
  return false
end function

sub initlex(n as integer)
  curline = n
  textp = 1
  initlex2()
end sub

sub initlex2()
  thelin = pgm(curline)
  thech = " "
  nexttok()
end sub

sub skiptoeol()
  tok = "": toktype = ""
  textp = len(thelin) + 1
end sub

sub nexttok()
  tok = "": toktype = ""
  begin:
  if thech <> "" then
    select case asc(thech)
      case is <= 32: getch(): goto begin
      case asc("a") to asc("z"), asc("A") to asc("Z"): readident(): if tok = "rem" then skiptoeol()
      case asc("0") to asc("9"): readint()
      case 34: readstr()             ' double quote
      case asc("'"): skiptoeol()
      case asc("#"),asc("("),asc(")"),asc("*"),asc("+"),asc(","),asc("-"),asc("/"),asc(":"),asc(";"),asc("<"),asc("="),asc(">"),asc("?"),asc("@"),asc("\"),asc("^")
        toktype = "punct"
        tok = thech
        getch()
        if tok = "<" or tok = ">" then
          if thech = "=" or thech = ">" then
            tok = tok + thech
            getch()
          end if
        end if
      case else
        toktype = ""
        writeline("What?" & thech & thelin): getch(): errors = true
    end select
  end if
end sub

' leave the " as the beginning of the string, so it won't get confused with other tokens
' especially in the print routines
sub readstr()
  tok = thech: toktype = "string"
  getch()
  do while thech <> chr(34)  ' while not a double quote
    if thech = "" then writeline("String not terminated"): errors = true: exit sub
    tok = tok + thech
    getch()
  loop
  getch()
end sub

sub readint()
  tok = "": toktype = "number"
  do while thech >= "0" and thech <= "9"
    tok = tok + thech
    getch()
  loop
  num = Convert.ToInt32(val(tok))
end sub

sub readident()
  tok = "": toktype = "ident"
  do while (thech >= "a" and thech <= "z") or (thech >= "A" and thech <= "Z")
    tok = tok + lcase(thech)
    getch()
  loop
end sub

sub getch()
  ' Any more text on this line?
  if textp > len(thelin) then thech = "": exit sub
  thech = mid(thelin, textp, 1)
  textp = textp + 1
end sub

end module

File: int-st.cs
// Ed Davis. Tiny Basic that can play Star Trek
// Supports: end, list, load, new, run, save
// gosub/return, goto, if, input, print, multi-statement lines (:)
// a single numeric array: @(n), and rnd(n)

using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using System.Timers;
using Microsoft.VisualBasic;

class Program {

const int c_maxlines = 7000, c_maxvars = 26, c_at_max = 1000;

static string[] pgm = new string[c_maxlines + 1]; // program stored here
static int[] vars = new int[c_maxvars + 1];       // variable store
static int[] atarry = new int[c_at_max];          // the @ array

static Stack<int> gstackln = new Stack<int>();
static Stack<int> gstacktp = new Stack<int>();

static int[] forvar = new int[c_maxvars];
static int[] forlimit = new int[c_maxvars];
static int[] forline = new int[c_maxvars];
static int[] forpos = new int[c_maxvars];

static string tok, toktype;                       // current token, and it's type
static string thelin, thech;                      // current program line, current character
static int curline, textp;                        // position in current line
static int num;                                   // last number read by scanner
static bool errors;
static bool tracing;

static TimeSpan timestart;
static Random random = new Random();

static void Main(string[] args) {
  newstmt();
  help();
  while (true) {
    errors = false;
    Console.Write("c#> ");
    pgm[0] = Console.ReadLine();
    if (pgm[0] != "") {
      initlex(0);
      if (toktype == "number") {
        validlinenum();
        pgm[num] = pgm[0].Substring(textp - 1);
      } else {
        docmd();
      }
    }
  }
}

static void docmd() {
  bool running = false;
  while (true) {
    if (tracing && tok[0] != ':') Console.WriteLine("[" + curline + "] " + tok + " " + thech + " " + thelin.Substring(textp - 1));
    switch (tok) {
      case "bye": case "quit": nexttok(); Environment.Exit(1); break;
      case "end": case "stop": nexttok();          return;
      case "clear"       : nexttok(); clearvars(); return;
      case "help"        : nexttok(); help();      return;
      case "list"        : nexttok(); liststmt();  return;
      case "load"        : nexttok(); loadstmt();  return;
      case "new"         : nexttok(); newstmt();   return;
      case "run"         : nexttok(); runstmt(); running = true; break;
      case "save"        : nexttok(); savestmt();  return;
      case "tron"        : nexttok(); tracing = true; break;
      case "troff"       : nexttok(); tracing = false; break;
      case "cls"         : nexttok(); Console.Clear(); break;
      case "for"         : nexttok(); forstmt();   break;
      case "gosub"       : nexttok(); gosubstmt(); break;
      case "goto"        : nexttok(); gotostmt();  break;
      case "if"          : nexttok(); ifstmt();    break;
      case "input"       : nexttok(); inputstmt(); break;
      case "next"        : nexttok(); nextstmt();  break;
      case "print": case "?":nexttok(); printstmt(); break;
      case "return"      : nexttok(); returnstmt();break;
      case "@"           : nexttok(); arrassn();   break;
      case ":"           : nexttok();              break;
      case ""            :                         break;
      default:
        if (tok == "let") nexttok();
        if (toktype == "ident") {
          assign();
        } else {
          Console.WriteLine("Unknown token " + tok + " at line " + curline); errors = true;
        }
        break;
    }

    if (errors) return;
    if (curline + 1 >= pgm.Length) {
      showtime(running);
      return;
    }
    while (tok == "") {
      if (curline == 0 || curline + 1 >= pgm.Length) {
        showtime(running);
        return;
      }
      initlex(curline + 1);
    }
  }
}

static void showtime(bool running) {
  TimeSpan timenow = (DateTime.UtcNow - new DateTime(1970, 1, 1));
  if (running)
    Console.WriteLine("Took : " + (timenow.TotalSeconds - timestart.TotalSeconds) + " seconds");
}

static void help() {
   Console.WriteLine("+---------------------- Tiny Basic Help (C#)--------------------------+" );
   Console.WriteLine("| bye, clear, end, help, list, load, new, run, save, stop             |");
   Console.WriteLine("| goto <expr>                                                         |");
   Console.WriteLine("| gosub <expr> ... return                                             |");
   Console.WriteLine("| if <expr> then <statement>                                          |");
   Console.WriteLine("| input [prompt,] <var>                                               |");
   Console.WriteLine("| <var>=<expr>                                                        |");
   Console.WriteLine("| print <expr|string>[,<expr|string>][;]                              |");
   Console.WriteLine("| rem <anystring>                                                     |");
   Console.WriteLine("| Operators: + - * / < <= > >= <> =                                   |");
   Console.WriteLine("| Integer variables a..z, and array @(expr)                           |");
   Console.WriteLine("| Functions: rnd(expr)                                                |");
   Console.WriteLine("+---------------------- Tiny Basic Help ------------------------------+" );
}

static void gosubstmt() {   // for gosub: save the line and column
  gstackln.Push(curline);
  gstacktp.Push(textp);

  gotostmt();
}

static void assign() {
  int var;
  var = getvarindex();
  nexttok();
  expect("=");
  vars[var] = expression(0);
  if (tracing) Console.WriteLine("*** " + (char)(var + 'a') + " = " + vars[var]);
}

static void arrassn() {   // array assignment: @(expr) = expr
  int n, atndx;

  atndx = parenexpr();
  if (tok != "=") {
    Console.WriteLine("Array Assign: Expecting '=', found: " + tok);
    errors = true;
  } else {
    nexttok();     // skip the "="
    n = expression(0);
    atarry[atndx] = n;
    if (tracing) Console.WriteLine("*** @(" + atndx + ") = " + n);
  }
}

static void forstmt() { // for i = expr to expr
  int var, forndx, n;

  var = getvarindex();
  assign();
  // vars(var) has the value; var has the number value of the variable in 0..25
  forndx = var;
  forvar[forndx] = vars[var];
  if (tok != "to") {
    Console.WriteLine("For: Expecting 'to', found:" + tok); errors = true;
  } else {
    nexttok();
    n = expression(0);
    forlimit[forndx] = n;
    // need to store iter, limit, line, and col
    forline[forndx] = curline;
    if (tok == "") forpos[forndx] = textp; else forpos[forndx] = textp - 2;
    //forpos[forndx] textp; if (tok != "") forpos[forndx] -=2;
  }
}

static void ifstmt() {
  if (expression(0) == 0) {
    skiptoeol();
    return;
  }
  if (tok == "then") nexttok();      // "then" is optional
  if (toktype == "number") gotostmt();
}

static void inputstmt() {   // "input" [string ","] var
  int var;
  string st;
  if (toktype == "string") {
    Console.Write(tok.Substring(1));
    nexttok();
    expect(",");
  } else {
    Console.Write("? ");
  }
  var = getvarindex();
  nexttok();
  st = Console.ReadLine();
  if (st == "") st = "0";
  if (Char.IsDigit(st[0])) {
    vars[var] = Int32.Parse(st);
  } else {
    vars[var] = (int)(st[0]); // turn characters into their ascii value
  }
}

static void liststmt() {
  int i;
  for (i = 1; i < pgm.Length; i++) {
    if (pgm[i] != "") Console.WriteLine(i + " " + pgm[i]);
  }
  Console.WriteLine("");
}

static void loadstmt() {
  int n;
  string filename;
  StreamReader f;

  newstmt();
  filename = getfilename("Load");
  if (filename == "") return;

  f = new StreamReader(filename);
  n = 0;
  while (f.Peek() > 0) {
    pgm[0] = f.ReadLine();
    initlex(0);
    if (toktype == "number" && num > 0 && num <= pgm.Length) {
      pgm[num] = pgm[0].Substring(textp - 1);
      n = num;
    } else {
      n++;
      pgm[n] = pgm[0];
    }
  }
  f.Close();
  curline = 0;
}

static void newstmt() {
  int i;
  clearvars();
  for (i = 1; i < pgm.Length; i++) {
    pgm[i] = "";
  }
}

static void nextstmt() {
  int forndx;

  // tok needs to have the variable
  forndx = getvarindex();
  forvar[forndx] = forvar[forndx] + 1;
  vars[forndx] = forvar[forndx];
  if (tracing) Console.WriteLine("*** " + (forndx + 'a') + " = " + vars[forndx]);
  if (forvar[forndx] <= forlimit[forndx]) {
    curline = forline[forndx];
    textp   = forpos[forndx];
    //print "nextstmt tok>"; tok; " textp>"; textp; " >"; mid$(thelin, textp)
    initlex2();
  } else {
    nexttok(); //' skip the ident for now
  }
}

// "print" expr { "," expr }] [","] {":" stmt} eol
// expr can also be a literal string
static void printstmt() {
  int printwidth;
  string junk;
  bool printnl = true;

  while (tok != ":" && tok != "") {
    printnl = true;
    printwidth = 0;
    if (accept("#")) {
      if (num <= 0) { Console.WriteLine("Expecting a print width, found:" + tok); return; }
      printwidth = num;
      nexttok();
      if (!accept(",")) { Console.WriteLine("Print: Expecting a ',', found:" + tok); return; }
    }

    if (toktype == "string") {
      junk = tok.Substring(1);
      nexttok();
    } else {
      junk = expression(0).ToString();
    }

    printwidth = printwidth - junk.Length;
    if (printwidth <= 0) { Console.Write(junk);} else {Console.Write(" ".PadRight(printwidth) + junk); }

    if (accept(",") || accept(";")) {printnl = false;} else {break; }
  }

  if (printnl) Console.WriteLine("");
}

static void returnstmt() {    // return from a subroutine
  curline = gstackln.Pop();
  textp   = gstacktp.Pop();
  initlex2();
}

static void runstmt() {
  timestart = (DateTime.UtcNow - new DateTime(1970, 1, 1));
  clearvars();
  initlex(1);
}

static void gotostmt() {
  num = expression(0);
  validlinenum();
  initlex(num);
}

static void savestmt() {
  int i;
  string filename;
  StreamWriter f;

  filename = getfilename("Save");
  if (filename == "") return;
  f = new StreamWriter(filename, false);
  for (i = 1; i < pgm.Length; i++) {
    if (pgm[i] != "") f.WriteLine(i + " " + pgm[i]);
  }
  f.Close();
}

static string getfilename(string action) {
  string filename;
  if (toktype == "string") {
    filename = tok.Substring(1);
  } else {
    Console.Write(action + ": ");
    filename = Console.ReadLine();
  }
  if (filename == "") return "";
  if (filename.IndexOf(".") == -1) filename += ".bas";
  return filename;
}

static void validlinenum() {
  if (num <= 0 || num >= pgm.Length) {
    Console.WriteLine("Line number out of range");
    errors = true;
  }
}

static void clearvars() {
  int i;
  for(i = 0; i < c_maxvars; i++) {
    vars[i] = 0;
  }
  gstackln.Clear();
  gstacktp.Clear();
}

static int parenexpr() {
  int n;

  expect("(");
  if (errors) return 0;
  n = expression(0);
  expect(")");
  return n;
}

static int expression(int minprec) {
  int n;

  // handle numeric operands - numbers and unary operators
  if (toktype == "number") { n = num; nexttok();
  } else if (tok == "-")   { nexttok(); n = -expression(7);
  } else if (tok == "+")   { nexttok(); n =  expression(7);
  } else if (tok == "not") { nexttok(); n = expression(3); if (n != 0) n = 1;
  } else if (tok == "abs") { nexttok(); n = Math.Abs(parenexpr());
  } else if (tok == "asc") { nexttok(); expect("("); n = tok[1]; nexttok(); expect(")");
  } else if (tok == "rnd" || tok == "irnd") { nexttok(); n = random.Next(1, parenexpr());
  } else if (tok == "sgn") { nexttok(); n = Math.Sign(parenexpr());
  } else if (toktype == "ident") { n = vars[getvarindex()]; nexttok();
  } else if (tok == "@") { nexttok(); n = atarry[parenexpr()];
  } else if (tok == "(") { n =  parenexpr();
  } else {
    Console.WriteLine("syntax error: expecting an operand, found: " + tok);
    errors = true;
    return 0;
  }

  while (true) { // while binary operator and precedence of tok >= minprec
    if (minprec <= 1 && tok == "or") { nexttok(); n = n | expression(2);
    } else if (minprec <= 2 && tok == "and") { nexttok(); n = n & expression(3);
    } else if (minprec <= 4 && tok == "=" )  { nexttok(); n = Convert.ToInt32(n == expression(5));
    } else if (minprec <= 4 && tok == "<" )  { nexttok(); n = Convert.ToInt32(n <  expression(5));
    } else if (minprec <= 4 && tok == ">" )  { nexttok(); n = Convert.ToInt32(n >  expression(5));
    } else if (minprec <= 4 && tok == "<>")  { nexttok(); n = Convert.ToInt32(n != expression(5));
    } else if (minprec <= 4 && tok == "<=")  { nexttok(); n = Convert.ToInt32(n <= expression(5));
    } else if (minprec <= 4 && tok == ">=")  { nexttok(); n = Convert.ToInt32(n >= expression(5));
    } else if (minprec <= 5 && tok == "+" )  { nexttok(); n = n + expression(6);
    } else if (minprec <= 5 && tok == "-" )  { nexttok(); n = n - expression(6);
    } else if (minprec <= 6 && tok == "*" )  { nexttok(); n = n * expression(7);
    } else if (minprec <= 6 && (tok == "/" || tok == "\\")) { nexttok(); n = n / expression(7);
    } else if (minprec <= 6 && tok == "mod") { nexttok(); n = n % expression(7);
    } else if (minprec <= 8 && tok == "^") { nexttok(); n = (int)Math.Pow(n, expression(9));
    } else { break; }
  }
  return n;
}

static int getvarindex() {
  if (toktype != "ident") {
    Console.WriteLine("Not a variable:" + tok);
    errors = true;
    return 0;
  }
  return tok[0] - 'a';
}

static void expect(string s) {
  if (accept(s)) return;
  Console.WriteLine("(" + curline + ") expecting " + s + " but found " + tok + " =>" + pgm[curline]);
  errors = true;
}

static bool accept(string s) {
  if (tok == s) {
    nexttok();
    return true;
  }
  return false;
}

static void initlex(int n) {
  curline = n;
  textp = 1;
  initlex2();
}

static void initlex2() {
  thelin = pgm[curline];
  thech = " ";
  nexttok();
}

static void skiptoeol() {
  tok = ""; toktype = "";
  textp = thelin.Length + 1;
}

static void nexttok() {
  tok = ""; toktype = "";
  for (;;) {
    if (thech == "") return;
    if (thech[0] > ' ') break;
    getch();
  }

  tok = thech;
  if (Char.IsLetter(tok[0])) {
    readident();
    if (tok == "rem") skiptoeol();
  } else if (Char.IsDigit(tok[0])) {
    readint();
  } else if (tok == "'") {
    skiptoeol();
  } else if (tok == "\"") {    // double quote - sstring
    readstr();
  } else {
    if ("#()*+,-/:;<=>?@\\^".IndexOf(tok) >= 0) {
      toktype = "punct";
      getch();
      if (tok == "<" || tok == ">") {
        if (thech == "=" || thech == ">") {
          tok = tok + thech;
          getch();
        }
      }
    } else {
      Console.WriteLine("What?" + thech + thelin);
      getch();
      errors = true;
    }
  }
}

// leave the " as the beginning of the string, so it won't get confused with other tokens
// especially in the print routines
static void readstr() {
  toktype = "string";
  getch();
  while (thech != "\"") {  // while not a double quote
    if (thech == "") {
      Console.WriteLine("String not terminated");
      errors = true;
      return;
    }
    tok = tok + thech;
    getch();
  }
  getch();
}

static void readint() {
  tok = ""; toktype = "number";
  while (thech != "" && Char.IsDigit(thech[0])) {
    tok += thech;
    getch();
  }
  num = Int32.Parse(tok);
}

static void readident() {
  tok = ""; toktype = "ident";
  while (thech != "" && Char.IsLetter(thech[0])) {
    tok = tok + thech.ToLower();
    getch();
  }
}

static void getch() {  // Any more text on this line?
  if (textp > thelin.Length) {
    thech = "";
    return;
  }
  thech = thelin[textp - 1].ToString();
  textp++;
}

}
